marimo/_ai/llm/_impl.py

```diff
@@ -664,3 +702,411 @@ def __call__(
+        from pydantic_ai.models.openai import OpenAIChatModel
+        from pydantic_ai.providers.openai import OpenAIProvider
+        from pydantic_ai.profiles.openai import OpenAIModelProfile
+
+        # Configure model for W&B Inference
+        model = OpenAIChatModel(
+            model_name="deepseek-ai/DeepSeek-R1-0528",
+            provider=OpenAIProvider(
+                api_key="your-wandb-key",
+                base_url="https://api.inference.wandb.ai/v1",
+            ),
+            profile=OpenAIModelProfile(
+                openai_chat_thinking_field="reasoning_content",
+            ),
+        )
+
+        agent = Agent(model, instructions="Think step by step.")
+        chat = mo.ui.chat(mo.ai.llm.pydantic_ai(agent))
+        ```
+    """
+
+    def __init__(self, agent: Any):
+        # Validate that we received an Agent-like object
+        if not hasattr(agent, "run_stream_events"):
+            raise TypeError(
+                "pydantic_ai() requires a pydantic_ai.Agent instance. "
+                "Create an Agent first: "
+                "Agent('openai:gpt-4.1', tools=[...], instructions='...')"
+            )
+        self._agent = agent
+
+    def __call__(
+        self, messages: list[ChatMessage], config: ChatModelConfig
+    ) -> AsyncGenerator[Any, None]:
+        """Returns an async generator that handles streaming and tool calls."""
+        DependencyManager.pydantic_ai.require(
+            "pydantic_ai chat model requires pydantic-ai. `pip install pydantic-ai`"
+        )
+        return self._stream_response(messages, config)
+
+    async def _stream_response(
+        self, messages: list[ChatMessage], config: ChatModelConfig
+    ) -> AsyncGenerator[Any, None]:
+        """Stream response from Pydantic AI agent.
+
+        Uses run_stream_events() to get real-time events for thinking,
+        text, and tool calls. Translates Pydantic AI events to marimo's
+        streaming format.
+        """
+        from pydantic_ai.messages import (
+            ModelMessagesTypeAdapter,
+            ModelRequest,
+            ModelResponse,
+            ThinkingPart,
+            ToolCallPart,
+            ToolReturnPart,
+            UserPromptPart,
+        )
+
+        del config  # Agent has its own model_settings
+
+        # Extract message history from previous assistant messages
+        # Pydantic AI stores complete history in all_messages_json()
+        message_history = self._get_message_history(
+            messages[:-1], ModelMessagesTypeAdapter
+        )
+
+        # Get the current user prompt
+        user_prompt = str(messages[-1].content) if messages else ""
+
+        # Track state for structured response
+        current_text = ""
+        current_thinking = ""
+        has_tool_calls = False
+        has_thinking = False
+        final_result: Any = None
+        pending_tool_calls: dict[str, dict[str, Any]] = {}
+
+        def _build_parts() -> list[dict[str, Any]]:
```

Comment:
@Light2Dark is adding some features now that should be able to remove this. either we can merge this as-is and delete later or wait (just a few days)